pkg/sql/sem/builtins/builtins.go

```diff
@@ -2537,6 +2552,90 @@ var jsonBuildArrayImpl = tree.Builtin{
 	Info: "Builds a possibly-heterogeneously-typed JSON or JSONB array out of a variadic argument list.",
 }
 
+var jsonObjectImpls = []tree.Builtin{
+	{
+		Types:      tree.ArgTypes{{"texts", types.TArray{Typ: types.String}}},
+		ReturnType: tree.FixedReturnType(types.JSON),
+		Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+			arr := tree.MustBeDArray(args[0])
+			if arr.Len()%2 != 0 {
+				return nil, errJSONObjectNotEvenNumberOfElements
+			}
+			builder := json.NewBuilder()
+			for i := 0; i < arr.Len(); i += 2 {
+				key, err := asJSONObjectKey(arr.Array[i])
+				if err != nil {
+					return nil, err
+				}
+				val, err := asJSON(arr.Array[i+1])
+				if err != nil {
+					return nil, err
+				}
+				builder.Add(key, val)
+			}
+			return tree.NewDJSON(builder.Build()), nil
+		},
+		Info: "Builds a JSON or JSONB object out of a text array. The array must have " +
+			"exactly one dimension with an even number of members, in which case " +
+			"they are taken as alternating key/value pairs.",
+	},
+	{
```

Comment:
I think we don't officially support arrays-of-arrays at this time, so I would be a little more comfortable if we left this one out for the time being, I don't really know the implications of arrays-of-arrays with the current codebase.